subsys/bluetooth/services/hids_c.c

```diff
@@ -213,6 +213,342 @@ static u16_t chrc_value_handle_by_uuid(struct bt_gatt_dm *dm,
 	return gatt_desc->handle;
 }
 
+/**
+ * @brief Mark hids ready to work
+ *
+ * This function marks ready flags and calls required callbacks.
+ * It releases parameter read semaphore.
+ *
+ * @param hids_c HIDS client object.
+ */
+static void hids_mark_ready(struct bt_gatt_hids_c *hids_c)
+{
+	k_sem_give(&hids_c->read_params_sem);
+	hids_c->ready = true;
+	if (hids_c->ready_cb) {
+		hids_c->ready_cb(hids_c);
+	}
+}
+
+/**
+ * @brief Call error callback
+ *
+ * This function calls HIDS client callback for error.
+ * It releases parameter read semaphore.
+ *
+ * @param hids_c HIDS client object.
+ * @param err    Error code.
+ */
+static void hids_prep_error(struct bt_gatt_hids_c *hids_c, int err)
+{
+	k_sem_give(&hids_c->read_params_sem);
+	if (hids_c->prep_error_cb) {
+		hids_c->prep_error_cb(hids_c, err);
+	}
+}
+
+/**
+ * @brief Process protocol mode read
+ *
+ * This function processed protocol mode read that is started when
+ * HID server is connected.
+ *
+ * @param conn   Connection handler.
+ * @param err    Read ATT error code.
+ * @param params Notification parameters structure - the pointer
+ *               to the structure provided to read function.
+ * @param data   Pointer to the data buffer.
+ * @param length The size of the received data.
+ *
+ * @retval BT_GATT_ITER_STOP     Stop notification
+ * @retval BT_GATT_ITER_CONTINUE Continue notification
+ */
+static u8_t pm_read_process(struct bt_conn *conn, u8_t err,
+			    struct bt_gatt_read_params *params,
+			    const void *data, u16_t length);
+
+/**
+ * @brief Start protocol mode read
+ *
+ * Function that starts protocol mode read at the HID connection.
+ * @note
+ * Read semaphore should be already taken in @ref repref_read_start.
+ *
+ * @param hids_c  See @ref bt_gatt_hids_c_handles_assign.
+ *
+ * @return 0 or negative error value.
+ */
+static int pm_read_start(struct bt_gatt_hids_c *hids_c)
+{
+	int err;
+
+	__ASSERT_NO_MSG(hids_c);
+	if (hids_c->handlers.pm == 0) {
+		LOG_DBG("Device ready without boot protocol");
+		hids_mark_ready(hids_c);
+		return 0;
+	}
+	LOG_DBG("PM read start");
+	hids_c->read_params.func = pm_read_process;
+	hids_c->read_params.handle_count  = 1;
+	hids_c->read_params.single.handle = hids_c->handlers.pm;
+	hids_c->read_params.single.offset = 0;
+	err = bt_gatt_read(hids_c->conn, &(hids_c->read_params));
+	if (err) {
+		LOG_ERR("PM read error (err: %d)", err);
+		return err;
+	}
+	return 0;
+}
+
+static u8_t pm_read_process(struct bt_conn *conn, u8_t err,
```

Comment:
Could you move it above the previous function, so that declaration for ```pm_read_process``` is not needed

Reply:
You can do it for all other cases

Reply:
I have ordered the implementation in the order it is executed to make it easy to read. Also note that repref_read_process calls repref_read_start internally so it would require repref_read_start declaration anyway. If you stay in the opinion that this is better I will change the order. Just take into account the description here before you decide.

Reply:
let's leave it as it is now

(if the declaration for repref_read_process  is necessary anyway, then let's make other process functions look the same)